### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [28]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)
student_data = student_data.rename(columns={"Student ID": "student_id"})

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Calculate the percentage of students who passed math **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [29]:
school_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   School ID    15 non-null     int64 
 1   school_name  15 non-null     object
 2   type         15 non-null     object
 3   size         15 non-null     int64 
 4   budget       15 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 728.0+ bytes


In [30]:
student_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39170 entries, 0 to 39169
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   student_id     39170 non-null  int64 
 1   student_name   39170 non-null  object
 2   gender         39170 non-null  object
 3   grade          39170 non-null  object
 4   school_name    39170 non-null  object
 5   reading_score  39170 non-null  int64 
 6   math_score     39170 non-null  int64 
dtypes: int64(3), object(4)
memory usage: 2.1+ MB


In [31]:
total_schools = school_data.school_name.nunique()
total_schools

15

In [32]:
total_students = len(student_data)
total_students

39170

In [33]:
#Total School Budget
total_budget = school_data.budget.sum()
total_budget

24649428

In [34]:
#Average Math Score
#Average Reading score
student_data.describe()

student_id  reading_score    math_score
count  39170.000000    39170.00000  39170.000000
mean   19584.500000       81.87784     78.985371
std    11307.549359       10.23958     12.309968
min        0.000000       63.00000     55.000000
25%     9792.250000       73.00000     69.000000
50%    19584.500000       82.00000     79.000000
75%    29376.750000       91.00000     89.000000
max    39169.000000       99.00000     99.000000

In [35]:
avg_read = student_data.reading_score.mean()
avg_read

81.87784018381414

In [36]:
avg_math = student_data.math_score.mean()
avg_math

78.98537145774827

In [37]:
df1 = student_data["reading_score"]>=70
df1 = df1.sum()
df2 = student_data["student_name"].notnull()
df2 = df2.sum()
read_pass = df1/df2*100
read_pass

85.80546336482001

In [38]:
df3 = student_data["math_score"]>=70
df3 = df3.sum()
math_pass = df3/df2*100
math_pass

74.9808526933878

In [39]:
df4 = (student_data["math_score"]>=70) & (student_data["reading_score"]>=70)
df4 = df4.sum()
both_pass = df4/df2*100
both_pass

65.17232575950983

In [40]:
#Create a dataframe to hold the above results
output = {
    "Total Schools": total_schools,
    "Total Students": total_students,
    "Total Budget": total_budget,
    'Average Math Score': avg_math,
    "Average Reading Score": avg_read,
    "% Passing Math": math_pass,
    "% Passing Reading": read_pass,
    "% Overall Passing": both_pass   
}
summary =pd.DataFrame([output])
summary

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

In [ ]:
#school_data_complete.rename()

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)
  
* Create a dataframe to hold the above results

In [52]:
#School Type & School Name
school_type = school_data.set_index(["school_name"])["type"]
school_type

school_name
Huang High School        District
Figueroa High School     District
Shelton High School       Charter
Hernandez High School    District
Griffin High School       Charter
Wilson High School        Charter
Cabrera High School       Charter
Bailey High School       District
Holden High School        Charter
Pena High School          Charter
Wright High School        Charter
Rodriguez High School    District
Johnson High School      District
Ford High School         District
Thomas High School        Charter
Name: type, dtype: object

In [116]:
#Total Students
total_stu = school_data_complete.groupby(["school_name", "type"]).student_id.count()
total_stu

school_name            type    
Bailey High School     District    4976
Cabrera High School    Charter     1858
Figueroa High School   District    2949
Ford High School       District    2739
Griffin High School    Charter     1468
Hernandez High School  District    4635
Holden High School     Charter      427
Huang High School      District    2917
Johnson High School    District    4761
Pena High School       Charter      962
Rodriguez High School  District    3999
Shelton High School    Charter     1761
Thomas High School     Charter     1635
Wilson High School     Charter     2283
Wright High School     Charter     1800
Name: student_id, dtype: int64

In [119]:
#Total School Budget
total_bud = school_data_complete.groupby(["school_name","type"]).mean()["budget"]
total_bud

school_name            type    
Bailey High School     District    3124928.0
Cabrera High School    Charter     1081356.0
Figueroa High School   District    1884411.0
Ford High School       District    1763916.0
Griffin High School    Charter      917500.0
Hernandez High School  District    3022020.0
Holden High School     Charter      248087.0
Huang High School      District    1910635.0
Johnson High School    District    3094650.0
Pena High School       Charter      585858.0
Rodriguez High School  District    2547363.0
Shelton High School    Charter     1056600.0
Thomas High School     Charter     1043130.0
Wilson High School     Charter     1319574.0
Wright High School     Charter     1049400.0
Name: budget, dtype: float64

In [121]:
#Per Student Budget
per_stu_bud = total_bud/total_stu
per_stu_bud

school_name            type    
Bailey High School     District    628.0
Cabrera High School    Charter     582.0
Figueroa High School   District    639.0
Ford High School       District    644.0
Griffin High School    Charter     625.0
Hernandez High School  District    652.0
Holden High School     Charter     581.0
Huang High School      District    655.0
Johnson High School    District    650.0
Pena High School       Charter     609.0
Rodriguez High School  District    637.0
Shelton High School    Charter     600.0
Thomas High School     Charter     638.0
Wilson High School     Charter     578.0
Wright High School     Charter     583.0
dtype: float64

In [122]:
#Average Math Score
avg_math = school_data_complete.groupby(["school_name","type"]).mean()["math_score"]
avg_math

school_name            type    
Bailey High School     District    77.048432
Cabrera High School    Charter     83.061895
Figueroa High School   District    76.711767
Ford High School       District    77.102592
Griffin High School    Charter     83.351499
Hernandez High School  District    77.289752
Holden High School     Charter     83.803279
Huang High School      District    76.629414
Johnson High School    District    77.072464
Pena High School       Charter     83.839917
Rodriguez High School  District    76.842711
Shelton High School    Charter     83.359455
Thomas High School     Charter     83.418349
Wilson High School     Charter     83.274201
Wright High School     Charter     83.682222
Name: math_score, dtype: float64

In [123]:
#Average Read Score
avg_read = school_data_complete.groupby(["school_name","type"]).mean()["reading_score"]
avg_read

school_name            type    
Bailey High School     District    81.033963
Cabrera High School    Charter     83.975780
Figueroa High School   District    81.158020
Ford High School       District    80.746258
Griffin High School    Charter     83.816757
Hernandez High School  District    80.934412
Holden High School     Charter     83.814988
Huang High School      District    81.182722
Johnson High School    District    80.966394
Pena High School       Charter     84.044699
Rodriguez High School  District    80.744686
Shelton High School    Charter     83.725724
Thomas High School     Charter     83.848930
Wilson High School     Charter     83.989488
Wright High School     Charter     83.955000
Name: reading_score, dtype: float64

In [124]:
#passing math
passed_math = school_data_complete.loc[school_data_complete.math_score >= 70].groupby(["school_name", "type"]).student_id.count()
pm = passed_math/total_stu

pm

school_name            type    
Bailey High School     District    0.666801
Cabrera High School    Charter     0.941335
Figueroa High School   District    0.659885
Ford High School       District    0.683096
Griffin High School    Charter     0.933924
Hernandez High School  District    0.667530
Holden High School     Charter     0.925059
Huang High School      District    0.656839
Johnson High School    District    0.660576
Pena High School       Charter     0.945946
Rodriguez High School  District    0.663666
Shelton High School    Charter     0.938671
Thomas High School     Charter     0.932722
Wilson High School     Charter     0.938677
Wright High School     Charter     0.933333
Name: student_id, dtype: float64

In [125]:
#% Passing read
passed_read = school_data_complete.loc[school_data_complete.reading_score >= 70].groupby(["school_name", "type"]).student_id.count()
pr = passed_read/total_stu

pr

school_name            type    
Bailey High School     District    0.819333
Cabrera High School    Charter     0.970398
Figueroa High School   District    0.807392
Ford High School       District    0.792990
Griffin High School    Charter     0.971390
Hernandez High School  District    0.808630
Holden High School     Charter     0.962529
Huang High School      District    0.813164
Johnson High School    District    0.812224
Pena High School       Charter     0.959459
Rodriguez High School  District    0.802201
Shelton High School    Charter     0.958546
Thomas High School     Charter     0.973089
Wilson High School     Charter     0.965396
Wright High School     Charter     0.966111
Name: student_id, dtype: float64

In [126]:
#Overall Passing %
passed_both = school_data_complete.loc[(school_data_complete.reading_score >= 70) &(school_data_complete.math_score >= 70)].groupby(["school_name", "type"]).student_id.count()
pb = passed_both/total_stu

pb

school_name            type    
Bailey High School     District    0.546423
Cabrera High School    Charter     0.913348
Figueroa High School   District    0.532045
Ford High School       District    0.542899
Griffin High School    Charter     0.905995
Hernandez High School  District    0.535275
Holden High School     Charter     0.892272
Huang High School      District    0.535139
Johnson High School    District    0.535392
Pena High School       Charter     0.905405
Rodriguez High School  District    0.529882
Shelton High School    Charter     0.898921
Thomas High School     Charter     0.909480
Wilson High School     Charter     0.905826
Wright High School     Charter     0.903333
Name: student_id, dtype: float64

In [172]:
size = school_data_complete.groupby(["school_name","type"]).mean()["size"]
size

school_name            type    
Bailey High School     District    4976.0
Cabrera High School    Charter     1858.0
Figueroa High School   District    2949.0
Ford High School       District    2739.0
Griffin High School    Charter     1468.0
Hernandez High School  District    4635.0
Holden High School     Charter      427.0
Huang High School      District    2917.0
Johnson High School    District    4761.0
Pena High School       Charter      962.0
Rodriguez High School  District    3999.0
Shelton High School    Charter     1761.0
Thomas High School     Charter     1635.0
Wilson High School     Charter     2283.0
Wright High School     Charter     1800.0
Name: size, dtype: float64

In [173]:
#Create a dataframe to hold the above results

df = pd.DataFrame()
df["Total Students"] = total_stu
df["Total Budget"] = total_bud
df["Per Student Budget"] = per_stu_bud
df["Average Math"] = avg_math
df["Average Reading"] = avg_read
df["% Passing Math"] = pm
df["% Passing Reading"] = pr
df["% Passing Both"] = pb
#added size to check work later on in bins
df["size"]= size

df

Total Students  Total Budget  \
school_name           type                                     
Bailey High School    District            4976     3124928.0   
Cabrera High School   Charter             1858     1081356.0   
Figueroa High School  District            2949     1884411.0   
Ford High School      District            2739     1763916.0   
Griffin High School   Charter             1468      917500.0   
Hernandez High School District            4635     3022020.0   
Holden High School    Charter              427      248087.0   
Huang High School     District            2917     1910635.0   
Johnson High School   District            4761     3094650.0   
Pena High School      Charter              962      585858.0   
Rodriguez High School District            3999     2547363.0   
Shelton High School   Charter             1761     1056600.0   
Thomas High School    Charter             1635     1043130.0   
Wilson High School    Charter             2283     1319574.0   
Wright High School    Charter             1800     1049400.0   

                                Per Student Budget  Average Math  \
school_name           type                                         
Bailey High School    District               628.0     77.048432   
Cabrera High School   Charter                582.0     83.061895   
Figueroa High School  District               639.0     76.711767   
Ford High School      District               644.0     77.102592   
Griffin High School   Charter                625.0     83.351499   
Hernandez High School District               652.0     77.289752   
Holden High School    Charter                581.0     83.803279   
Huang High School     District               655.0     76.629414   
Johnson High School   District               650.0     77.072464   
Pena High School      Charter                609.0     83.839917   
Rodriguez High School District               637.0     76.842711   
Shelton High School   Charter                600.0     83.359455   
Thomas High School    Charter                638.0     83.418349   
Wilson High School    Charter                578.0     83.274201   
Wright High School    Charter                583.0     83.682222   

                                Average Reading  % Passing Math  \
school_name           type                                        
Bailey High School    District        81.033963        0.666801   
Cabrera High School   Charter         83.975780        0.941335   
Figueroa High School  District        81.158020        0.659885   
Ford High School      District        80.746258        0.683096   
Griffin High School   Charter         83.816757        0.933924   
Hernandez High School District        80.934412        0.667530   
Holden High School    Charter         83.814988        0.925059   
Huang High School     District        81.182722        0.656839   
Johnson High School   District        80.966394        0.660576   
Pena High School      Charter         84.044699        0.945946   
Rodriguez High School District        80.744686        0.663666   
Shelton High School   Charter         83.725724        0.938671   
Thomas High School    Charter         83.848930        0.932722   
Wilson High School    Charter         83.989488        0.938677   
Wright High School    Charter         83.955000        0.933333   

                                % Passing Reading  % Passing Both    size  
school_name           type                                                 
Bailey High School    District           0.819333        0.546423  4976.0  
Cabrera High School   Charter            0.970398        0.913348  1858.0  
Figueroa High School  District           0.807392        0.532045  2949.0  
Ford High School      District           0.792990        0.542899  2739.0  
Griffin High School   Charter            0.971390        0.905995  1468.0  
Hernandez High School District           0.808630        0.535275  4635.0  
Holden High School    Charter         

#Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)

#highlight from right to left and then wrap

# use markdown to create notes within file

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [138]:
#Top Performing Schools
df =df.sort_values('% Passing Both')
df.tail(5)

Total Students  Total Budget  Per Student Budget  \
school_name         type                                                        
Pena High School    Charter             962      585858.0               609.0   
Wilson High School  Charter            2283     1319574.0               578.0   
Griffin High School Charter            1468      917500.0               625.0   
Thomas High School  Charter            1635     1043130.0               638.0   
Cabrera High School Charter            1858     1081356.0               582.0   

                             Average Math  Average Reading  % Passing Math  \
school_name         type                                                     
Pena High School    Charter     83.839917        84.044699        0.945946   
Wilson High School  Charter     83.274201        83.989488        0.938677   
Griffin High School Charter     83.351499        83.816757        0.933924   
Thomas High School  Charter     83.418349        83.848930        0.932722   
Cabrera High School Charter     83.061895        83.975780        0.941335   

                             % Passing Reading  % Passing Both  
school_name         type                                        
Pena High School    Charter           0.959459        0.905405  
Wilson High School  Charter           0.965396        0.905826  
Griffin High School Charter           0.971390        0.905995  
Thomas High School  Charter           0.973089        0.909480  
Cabrera High School Charter           0.970398        0.913348

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [136]:
#Bottom Performing Schools
#df =df.sort_values('% Passing Both')
df.head(5)


Total Students  Total Budget  \
school_name           type                                     
Rodriguez High School District            3999     2547363.0   
Figueroa High School  District            2949     1884411.0   
Huang High School     District            2917     1910635.0   
Hernandez High School District            4635     3022020.0   
Johnson High School   District            4761     3094650.0   

                                Per Student Budget  Average Math  \
school_name           type                                         
Rodriguez High School District               637.0     76.842711   
Figueroa High School  District               639.0     76.711767   
Huang High School     District               655.0     76.629414   
Hernandez High School District               652.0     77.289752   
Johnson High School   District               650.0     77.072464   

                                Average Reading  % Passing Math  \
school_name           type                                        
Rodriguez High School District        80.744686        0.663666   
Figueroa High School  District        81.158020        0.659885   
Huang High School     District        81.182722        0.656839   
Hernandez High School District        80.934412        0.667530   
Johnson High School   District        80.966394        0.660576   

                                % Passing Reading  % Passing Both  
school_name           type                                         
Rodriguez High School District           0.802201        0.529882  
Figueroa High School  District           0.807392        0.532045  
Huang High School     District           0.813164        0.535139  
Hernandez High School District           0.808630        0.535275  
Johnson High School   District           0.812224        0.535392

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [149]:
dt =school_data_complete.groupby(["school_name","grade"]).mean()["math_score"]
dt

school_name            grade
Bailey High School     10th     76.996772
                       11th     77.515588
                       12th     76.492218
                       9th      77.083676
Cabrera High School    10th     83.154506
                       11th     82.765560
                       12th     83.277487
                       9th      83.094697
Figueroa High School   10th     76.539974
                       11th     76.884344
                       12th     77.151369
                       9th      76.403037
Ford High School       10th     77.672316
                       11th     76.918058
                       12th     76.179963
                       9th      77.361345
Griffin High School    10th     84.229064
                       11th     83.842105
                       12th     83.356164
                       9th      82.044010
Hernandez High School  10th     77.337408
                       11th     77.136029
                       12th     77.186567
     

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [150]:
dt =school_data_complete.groupby(["school_name","grade"]).mean()["reading_score"]
dt

school_name            grade
Bailey High School     10th     80.907183
                       11th     80.945643
                       12th     80.912451
                       9th      81.303155
Cabrera High School    10th     84.253219
                       11th     83.788382
                       12th     84.287958
                       9th      83.676136
Figueroa High School   10th     81.408912
                       11th     80.640339
                       12th     81.384863
                       9th      81.198598
Ford High School       10th     81.262712
                       11th     80.403642
                       12th     80.662338
                       9th      80.632653
Griffin High School    10th     83.706897
                       11th     84.288089
                       12th     84.013699
                       9th      83.369193
Hernandez High School  10th     80.660147
                       11th     81.396140
                       12th     80.857143
     

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

average Spending Ranges based on avg budget per student
school_name	Budget per

    *Wilson High School	578
    *Holden High School	581
    *Cabrera High School	582
    *Wright High School	583

    Shelton High School	600
    Pena High School	609
    Griffin High School	625
    Bailey High School	628

    Rodriguez High School	637
    Thomas High School	638
    Figueroa High School	639
    Ford High School	644

    Johnson High School	650
    Hernandez High School	652
    Huang High School	655


In [186]:
#copy above data range
bins = [0, 584, 629, 645, 656]

groups = ["<$584", "$585-$629", "$630-$644", "$645-$656"]

df2_copy = df.copy()

df2_copy["budget_bin"] = pd.cut(df2_copy["Per Student Budget"], bins, labels=groups, include_lowest=True)

df2_copy.budget_bin.value_counts()

<$584        4
$585-$629    4
$630-$644    4
$645-$656    3
Name: budget_bin, dtype: int64

## Scores by School Size

* Perform the same operations as above, based on school size.

    school_name	size
    Holden High School	427
    Pena High School	962
    Griffin High School	1468
    Thomas High School	1635
    Shelton High School	1761
    Wright High School	1800
    Cabrera High School	1858
    Wilson High School	2283
    Ford High School	2739
    Huang High School	2917
    Figueroa High School	2949
    Rodriguez High School	3999
    Hernandez High School	4635
    Johnson High School	4761
    Bailey High School	4976


In [187]:
#copy above data range
bins2 = [0, 1636, 2285, 4000, 5000]

groups2 = ["<1,636", "1,637-2,284", "2,285-3,999", "4,000-5,000"]

df3_copy = df.copy()

df3_copy["size_bin"] = pd.cut(df3_copy["size"], bins2, labels=groups2, include_lowest=True)

df3_copy.size_bin.value_counts()

<1,636         4
1,637-2,284    4
2,285-3,999    4
4,000-5,000    3
Name: size_bin, dtype: int64

## Scores by School Type

* Perform the same operations as above, based on school type

Charter vs. District

In [185]:
school_data_complete.type.value_counts()

District    26976
Charter     12194
Name: type, dtype: int64